versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/DatabaseAdapterUtil.java

```diff
@@ -0,0 +1,145 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.adapter.spi;
+
+import com.google.common.hash.Hasher;
+import com.google.common.hash.Hashing;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import org.projectnessie.versioned.BranchName;
+import org.projectnessie.versioned.Hash;
+import org.projectnessie.versioned.Key;
+import org.projectnessie.versioned.NamedRef;
+import org.projectnessie.versioned.ReferenceAlreadyExistsException;
+import org.projectnessie.versioned.ReferenceConflictException;
+import org.projectnessie.versioned.ReferenceNotFoundException;
+import org.projectnessie.versioned.persist.adapter.DatabaseAdapter;
+
+/**
+ * Utility methods for {@link DatabaseAdapter} implementations.
+ *
+ * <p>This class is just there to keep {@link AbstractDatabaseAdapter} cleaner and shorter.
+ */
+public abstract class DatabaseAdapterUtil implements DatabaseAdapter {
+
+  @SuppressWarnings("UnstableApiUsage")
+  protected static Hasher newHasher() {
+    return Hashing.sha256().newHasher();
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  protected static void hashKey(Hasher hasher, Key k) {
+    k.getElements().forEach(e -> hasher.putString(e, StandardCharsets.UTF_8));
+  }
+
+  protected static ReferenceConflictException hashCollisionDetected() {
+    return new ReferenceConflictException("Hash collision detected");
+  }
+
+  /** Builds a {@link ReferenceNotFoundException} exception with a human-readable message. */
+  protected static ReferenceNotFoundException hashNotFound(NamedRef ref, Hash hash) {
+    return new ReferenceNotFoundException(
+        String.format(
+            "Could not find commit '%s' in reference '%s'.", hash.asString(), ref.getName()));
+  }
+
+  protected static ReferenceNotFoundException refNotFound(NamedRef ref) {
+    return new ReferenceNotFoundException(
+        String.format("Named reference '%s' not found", ref.getName()));
+  }
+
+  protected static ReferenceAlreadyExistsException referenceAlreadyExists(NamedRef ref) {
+    return new ReferenceAlreadyExistsException(
+        String.format("Named reference '%s' already exists.", ref));
+  }
+
+  /** Returns the microseconds since epoch. */
+  protected static long currentTimeInMicros() {
+    // We only have System.currentTimeMillis() as the current wall-clock-value, but want
+    // microsecond "precision" here, which is implemented by taking the microsecond-part from
+    // System.nanoTime().
+    long microsFraction = TimeUnit.NANOSECONDS.toMicros(System.nanoTime());
```

Comment:
we can't rely on nanotime being fixed to a second so this is going to give us some microseconds since some completely arbitrary time. Why would we need such precision? If we do I would rather introduce a `Clock` and let more sophisticated clocks get injected. Finally can we get the time from `java.time` objects and make sure its in UTC

Reply:
Changed to `Clock`, but we do not need the microsecond part to be accurate.
Maybe we do not even need microsecond precision here (just took what's been used for that "dt" field in the current TVS).
All we basically need is some wall clock that tells us, whether a commit is "too recent" and "eligible" for Nessie-GC.